File: DesignPatternsExercises/java/VisitorDesignPattern/Solution/GeneralReport.java
public class GeneralReport implements Visitor {
    private int customersNumber;
    private int ordersNumber;

    @Override
    // public void visit(__________(4)__________ customer) {
    public void visit(Customer customer) {
        System.out.println("Customer: " + customer.getName());
        customersNumber++;
    }

    @Override
    // public void visit(__________(5)__________ order) {
    public void visit(Order order) {
        System.out.println("    " + order.getName());
        System.out.println("    Items:");
        for (String item : order.getItems())
            System.out.println("        " + item);
        ordersNumber++;
    }

    // public void __________(6)__________() {
    public void displayResults() {
        System.out.println("Number of customers: " + customersNumber);
        System.out.println("Number of orders:    " + ordersNumber);
    }
}

File: DesignPatternsExercises/ruby/visitor_design_pattern/solution/general_report.rb
require_relative 'visitor'
require_relative 'customer'

class GeneralReport < Visitor
  # def __________(6)__________
  def visit(node)
    # if node.respond_to?(__________(7)__________)
    if node.respond_to?(:items)
      puts "    #{node.name}"
      puts '    Items:'
      # node.items.__________(8)__________ { |order| puts "        #{order}" }
      node.items.each { |order| puts "        #{order}" }
      @orders_number ||= 0
      @orders_number += 1
    else
      puts "Customer: #{node.name}"
      @customers_number ||= 0
      @customers_number += 1
    end
  end

  def display_results
    puts "Number of customers: #{@customers_number}"
    puts "Number of orders:    #{@orders_number}"
  end
end
